esp/esp/program/controllers/autoscheduler/controller.py

```diff
@@ -0,0 +1,394 @@
+            section.grade_min, section.grade_max))
+        resources = ""
+        for restype in section.resource_requests.itervalues():
+            resources += "<li>"
+            if restype.value == "":
+                resources += restype.name
+            else:
+                resources += u"{}: {}".format(restype.name, restype.value)
+            resources += "</li>"
+        info.append(u"<b>Resource requests:</b><ul>{}</ul>".format(
+            resources))
+        cls = ClassSection.objects.get(id=section.id).parent_class
+        info.append(u"<b>Class Flags: </b>{}".format(u", ".join(
+            cls.flags.values_list(
+                'flag_type__name', flat=True))))
+        return info
+
+    def roomslot_identifier(self, roomslot):
+        return u"{} starting at {}".format(
+                roomslot.room.name,
+                roomslot.timeslot.start.strftime("%A %-I:%M%p"))
+
+    def roomslot_info(self, roomslot):
+        info = []
+        info.append(u"<b>Capacity:</b> {}".format(roomslot.room.capacity))
+        resources = ""
+        for restype in roomslot.room.furnishings.itervalues():
+            resources += "<li>"
+            if restype.value == "":
+                resources += restype.name
+            else:
+                resources += u"{}: {}".format(restype.name, restype.value)
+            resources += "</li>"
+        info.append(u"<b>Furnishings:</b><ul>{}</ul>".format(
+            resources))
+        return info
+
+    def simplify_history(self, history):
+        """Given a history object, this simplifies it. It is NOT necessarily an
+        equivalent object, but it should be more understandable, e.g. if you
+        unschedule and reschedule a class, it will count as a move."""
+        # Dict mapping from section to [original_roomslot, final_roomslot]
+        sections = {}
+        for action in history:
+            if action["action"] == "schedule":
+                section = action["section"]
+                if section not in sections:
+                    sections[section] = [None, action["start_roomslot"]]
+                else:
+                    sections[section][1] = action["start_roomslot"]
+            elif action["action"] == "move":
+                section = action["section"]
+                if section not in sections:
+                    sections[section] = [
+                        action["prev_start_roomslot"],
+                        action["start_roomslot"]]
+                else:
+                    sections[section][1] = action["start_roomslot"]
+            elif action["action"] == "unschedule":
+                section = action["section"]
+                if section not in sections:
+                    sections[section] = [action["prev_start_roomslot"], None]
+                else:
+                    sections[section][1] = None
+            elif action["action"] == "swap":
+                for section, old_r, new_r in zip(
```

Comment:
I'd like a comment here that you're iterating over exactly two sections in this loop (?)